Add MI300X tuning guide (#3372)
ROCm/ROCm: docs/how-to/tuning-guides/mi300x.rst

   Schematic representation of a CU in the CDNA2 or CDNA3 architecture.

The following is a list of kernel arguments used for tuning.

``num_stages=n``
   Adjusts the number of pipeline stages for different types of kernels. On AMD accelerators, set ``num_stages``
   according to the following rules:

   * For kernels with a single GEMM, set to ``0``.

   * For kernels with two GEMMs fused (Flash Attention, or any other kernel
     that fuses 2 GEMMs), set to ``1``.

   * For kernels that fuse a single GEMM with another non-GEMM operator
     (for example ReLU activation), set to ``0``.

   * For kernels that have no GEMMs, set to ``1``.

``waves_per_eu=n``
   Helps to manage Vector General Purpose Registers (VGPR) usage to achieve
   desired occupancy levels. This argument hints to the compiler to reduce VGPR
   to achieve ``n`` occupancy where ``n`` is a number. The goal is to achieve a
   certain occupancy level for each Execution Unit (EU, also called
   :ref:`SIMD Unit <mi300x-cu-fig>`) to achieve better latency or throughput.
   See :ref:`mi300x-compute-kernel-occ` for more information about how to
   compute occupancy. 

   This argument is useful if:

>    * The occupancy of the kernel is limited by VGPR usage.

Reviewer: ```suggestion
   * The occupancy of the kernel is limited by VGPR usage, and
```
Since the tune has changed from original documentation, please make sure we have the "and" here so that it means both conditions need to be met.

Author: Done

   * The current VGPR usage is only a few above a boundary in
     :ref:`Occupancy related to VGPR usage in an Instinct MI300X accelerator <mi300x-occupancy-vgpr-table>`.

.. _mi300x-occupancy-vgpr-table:

.. figure:: ../../data/shared/occupancy-vgpr.png
   :alt: Occupancy related to VGPR usage in an Instinct MI300X accelerator.
   :align: center

   Occupancy related to VGPRs usage in an Instinct MI300X accelerator

   For example, according to the table, the available VGPR is 512 per Execution
   Unit (EU), and VGPU is allocated at the unit of 16. If the current VGPR usage
   is 170, the actual requested VGPR will be 176, so the occupancy is only 2

>    waves per CU since :math:`176 \times 3 > 512`. So, if you set

Reviewer: ```suggestion
   waves per EU since :math:`176 \times 3 > 512`. So, if you set
```
Please accept this suggestion as we are talking about waves per EU. Original documentation is also updated. Thanks

Author: Done

   /app/vllm/benchmarks/benchmark_throughput.py --gpu-memory-utilization
   0.4 --dataset
   "/path/to/dataset/ShareGPT_V3_unfiltered_cleaned_split.json" --model
   /path/to/model &

Similarly, use the ``ROCR_VISIBLE_DEVICES`` environment variable to specify
which GPU (0-7) will run those instances. For example, the setting
``ROCR_VISIBLE_DEVICES="4,5,6,7"`` exposes GPUs 4, 5, 6, and 7 to the program.
Inside vLLM, the GPUs are mapped to ``CUDA_VISIBLE_DEVICES`` as 0, 1, 2, and 3.

Run vLLM on multiple GPUs
-------------------------

The two main reasons to use multiple GPUs:

*  The model size is too big to run vLLM using one GPU as it results
   CUDA/HIP Out of Memory.

*  To achieve better latency.

To run one vLLM instance on multiple GPUs, use the ``-tp`` or
``--tensor-parallel-size`` option to specify multiple GPUs. Optionally, use the
``ROCR_VISIBLE_DEVICES`` environment variable to specify the GPUs.

For example, we can use two GPUs to start an API server on port 8000 as
below:

.. code-block:: shell

   python -m vllm.entrypoints.api_server --model /path/to/model --dtype
   float16 -tp 2 --port 8000 &

If issues occur using multiple GPUs, specify the ``--enforce-eager``

> parameter.

Reviewer: We can remove this line, as the vllm is updated with latest fix and the flag is no longer needed.

Author: Done

predefined list of Triton configurations and selects the fastest one for each
shape. See the configurations in PyTorch source code:

* `conv configs for max-autotune <https://github.com/pytorch/pytorch/blob/a1d02b423c6b4ccacd25ebe86de43f650463bbc6/torch/_inductor/kernel/conv.py#L51>`_

* `matmul configs for max-autotune <https://github.com/pytorch/pytorch/blob/a1d02b423c6b4ccacd25ebe86de43f650463bbc6/torch/_inductor/kernel/mm_common.py#L118>`_

.. note::
   Triton is not used if regular :doc:`MIOpen <miopen:index>` or
   :doc:`rocBLAS <rocblas:index>` performs faster for a specific operation.

* Set ``torch._inductor.config.max_autotune = True`` or ``TORCHINDUCTOR_MAX_AUTOTUNE=1``.

* Or, for more fine-grained control:

  ``torch._inductor.config.max_autotune_gemm = True``
     To enable tuning or lowering of ``mm``/``conv``\s.

  ``torch._inductor.config.max_autotune.pointwise = True``
     To enable tuning for ``pointwise``/``reduction`` ops.

  ``torch._inductor.max_autotune_gemm_backends`` or ``TORCHINDUCTOR_MAX_AUTOTUNE_GEMM_BACKENDS``
     Selects the candidate backends for ``mm`` auto-tuning. Defaults to

>      ``TRITON,ATEN``. This also includes the ``CUTLASS`` tuning option for

Reviewer: Can you remove the sentence of "This also include the CUTLASS tuning option for NVIDIA CUDA". We don't need to mention NVIDIA in this document.

>      ``TRITON,ATEN``. This also includes the ``CUTLASS`` tuning option for

Reviewer: ```suggestion
     ``TRITON,ATEN``. 
```

>      NVIDIA CUDA. Limiting this to ``TRITON`` might improve performance by

Reviewer: ```suggestion
     Limiting this to ``TRITON`` might improve performance by
```

for additional performance tips.

Maximize throughput
-------------------

The general guideline is to maximize per-node throughput. Specify proper
GPU memory utilization to run as many instances of vLLM as possible on a
single GPU. However, too many instances can result in no memory for
KV-cache.

You can run vLLM on MI300X (gfx942), for example, using model weights
for ``llama2`` (``7b``, ``13b``, ``70b``) and ``llama3`` models (``8b``,
``70b``). 

As described in the
`AMD Instinct MI300X Accelerator <https://www.amd.com/content/dam/amd/en/documents/instinct-tech-docs/data-sheets/amd-instinct-mi300x-data-sheet.pdf>`__
data sheet, the GPU memory capacity is 192 GB. This means you can run
llama2-70b and llama3-70b models on one GPU.

To maximize the accumulated throughput, you can also run eight instances
vLLM simultaneously on one MI300X node (with eight GPUs). To do so, use
the GPU isolation environment variable ``ROCR_VISIBLE_DEVICES``.

For example, this script runs eight instances of vLLM for throughput
benchmarking at the same time:

.. code-block:: shell

   ROCR_VISIBLE_DEVICES="0" python3

>    /app/vllm/benchmarks/benchmark_throughput.py --dataset

Reviewer: ```suggestion
   /vllm-workspace/benchmarks/benchmark_throughput.py --dataset
```

   "/path/to/dataset/ShareGPT_V3_unfiltered_cleaned_split.json" --model
   /path/to/model &

   ROCR_VISIBLE_DEVICES="1" python3

>    /app/vllm/benchmarks/benchmark_throughput.py --dataset

Reviewer: ```suggestion
   /vllm-workspace/benchmarks/benchmark_throughput.py --dataset
```

   "/path/to/dataset/ShareGPT_V3_unfiltered_cleaned_split.json" --model
   /path/to/model &

   ROCR_VISIBLE_DEVICES="2" python3

>    /app/vllm/benchmarks/benchmark_throughput.py --dataset

Reviewer: ```suggestion
   /vllm-workspace/benchmarks/benchmark_throughput.py --dataset
```

   "/path/to/dataset/ShareGPT_V3_unfiltered_cleaned_split.json" --model
   /path/to/model &

   ROCR_VISIBLE_DEVICES="3" python3

>    /app/vllm/benchmarks/benchmark_throughput.py --dataset

Reviewer: ```suggestion
   /vllm-workspace/benchmarks/benchmark_throughput.py --dataset
```

   "/path/to/dataset/ShareGPT_V3_unfiltered_cleaned_split.json" --model
   /path/to/model &

   ROCR_VISIBLE_DEVICES="4" python3

>    /app/vllm/benchmarks/benchmark_throughput.py --dataset

Reviewer: ```suggestion
   /vllm-workspace/benchmarks/benchmark_throughput.py --dataset
```

   "/path/to/dataset/ShareGPT_V3_unfiltered_cleaned_split.json" --model
   /path/to/model &

   ROCR_VISIBLE_DEVICES="5" python3

>    /app/vllm/benchmarks/benchmark_throughput.py --dataset

Reviewer: ```suggestion
   /vllm-workspace/benchmarks/benchmark_throughput.py --dataset
```

   "/path/to/dataset/ShareGPT_V3_unfiltered_cleaned_split.json" --model
   /path/to/model &

   ROCR_VISIBLE_DEVICES="6" python3

>    /app/vllm/benchmarks/benchmark_throughput.py --dataset

Reviewer: ```suggestion
   /vllm-workspace/benchmarks/benchmark_throughput.py --dataset
```

   "/path/to/dataset/ShareGPT_V3_unfiltered_cleaned_split.json" --model
   /path/to/model &

   ROCR_VISIBLE_DEVICES="7" python3

>    /app/vllm/benchmarks/benchmark_throughput.py --dataset

Reviewer: ```suggestion
   /vllm-workspace/benchmarks/benchmark_throughput.py --dataset
```

   "/path/to/dataset/ShareGPT_V3_unfiltered_cleaned_split.json" --model
   /path/to/model &

Run two instances of ``llama3-8b`` model at the same time on one single GPU
by specifying ``--gpu-memory-utilization`` to 0.4 (40%), as below (on GPU
0):

.. code-block:: shell

   ROCR_VISIBLE_DEVICES=0 python3

>    /app/vllm/benchmarks/benchmark_throughput.py --gpu-memory-utilization

Reviewer: ```suggestion
   /vllm-workspace/benchmarks/benchmark_throughput.py --gpu-memory-utilization
```

   0.4 --dataset
   "/path/to/dataset/ShareGPT_V3_unfiltered_cleaned_split.json" --model
   /path/to/model &

   ROCR_VISIBLE_DEVICES=0 python3

>    /app/vllm/benchmarks/benchmark_throughput.py --gpu-memory-utilization

Reviewer: ```suggestion
   /vllm-workspace/benchmarks/benchmark_throughput.py --gpu-memory-utilization
```

   the Docker image without Flash Attention by passing ``--build-arg BUILD_FA="0"``
   during Docker build, or ``pip uninstall flash-attn``
   inside the container, and export ``VLLM_USE_TRITON_FLASH_ATTN=0`` when
   running the vLLM instance.

Use fp8 KV-cache data type
--------------------------

Using ``fp8 kv-cache dtype`` can improve performance as it reduces the size
of ``kv-cache``. As a result, it reduces the cost required for reading and
writing the ``kv-cache``.

To use this feature, specify ``--kv-cache-dtype`` as ``fp8``.

To specify the quantization scaling config, use the
``--quantization-param-path`` parameter. If the parameter isn’t specified,
the default scaling factor of ``1`` is used, which can lead to less accurate
results. To generate ``kv-cache`` scaling JSON file, see `FP8 KV
Cache <https://github.com/vllm-project/vllm/blob/main/examples/fp8/README.md>`__
in the vLLM GitHub repository.

Two sample Llama scaling configuration files are in vLLM for ``llama2-70b`` and
``llama2-7b``.

If building the vLLM using
`Dockerfile.rocm <https://github.com/vllm-project/vllm/blob/main/Dockerfile.rocm>`_
for ``llama2-70b`` scale config, find the file at

> ``/app/vllm/tests/fp8_kv/llama2-70b-fp8-kv/kv_cache_scales.json`` at

Reviewer: ```suggestion
``/vllm-workspace/tests/fp8_kv/llama2-70b-fp8-kv/kv_cache_scales.json`` at
```

runtime.

Below is a sample command to run benchmarking with this feature enabled
for the ``llama2-70b`` model:

.. code-block:: shell

>    python3 /app/vllm/benchmarks/benchmark_throughput.py --model

Reviewer: ```suggestion
   python3 /vllm-workspace/benchmarks/benchmark_throughput.py --model
```

   /path/to/llama2-70b-model --kv-cache-dtype "fp8"
   --quantization-param-path

>    "/app/vllm/tests/fp8_kv/llama2-70b-fp8-kv/kv_cache_scales.json"

Reviewer: ```suggestion
   "/vllm-workspace/tests/fp8_kv/llama2-70b-fp8-kv/kv_cache_scales.json"
```

predefined list of Triton configurations and selects the fastest one for each
shape. See the configurations in PyTorch source code:

* `conv configs for max-autotune <https://github.com/pytorch/pytorch/blob/a1d02b423c6b4ccacd25ebe86de43f650463bbc6/torch/_inductor/kernel/conv.py#L51>`_

* `matmul configs for max-autotune <https://github.com/pytorch/pytorch/blob/a1d02b423c6b4ccacd25ebe86de43f650463bbc6/torch/_inductor/kernel/mm_common.py#L118>`_

.. note::
   Triton is not used if regular :doc:`MIOpen <miopen:index>` or
   :doc:`rocBLAS <rocblas:index>` performs faster for a specific operation.

* Set ``torch._inductor.config.max_autotune = True`` or ``TORCHINDUCTOR_MAX_AUTOTUNE=1``.

* Or, for more fine-grained control:

  ``torch._inductor.config.max_autotune.pointwise = True``
     To enable tuning for ``pointwise``/``reduction`` ops.

  ``torch._inductor.config.max_autotune_gemm = True``
     To enable tuning or lowering of ``mm``/``conv``\s.

  ``torch._inductor.max_autotune_gemm_backends/TORCHINDUCTOR_MAX_AUTOTUNE_GEMM_BACKENDS``
     To select the candidate backends for ``mm`` auto-tuning. Defaults to

>      ``TRITON,ATEN,NV``. This also includes the ``CUTLASS`` tuning option. Limiting this to

Author: Done in [7948038](https://github.com/ROCm/ROCm/pull/3372/commits/794803898e80bca76e3a5c9fedf234b1e89ba4f1)

     ``TRITON`` might improve performance by enabling more fused ``mm`` kernels
     instead of going to rocBLAS.

* For further ``mm`` tuning, tuning ``coordinate_descent`` might improve
  performance.

  ``torch._inductor.config.coordinate_descent_tuning = True`` or ``TORCHINDUCTOR_COORDINATE_DESCENT_TUNING=1``

* Inference can see large improvements on AMD GPUs by utilizing
  ``torch._inductor.config.freezing=True`` or the ``TORCHINDUCTOR_FREEZING=1`` variable, which
  in-lines weights as constants and enables constant folding optimizations.

* Enabling ``inductor``’s cpp_wrapper might improve overhead. This generates
  C++ code which launches Triton binaries directly with
  ``hipModuleLaunchKernel`` and relies on `hipification`.

  ``torch._inductor.config.cpp_wrapper=True`` or ``TORCHINDUCTOR_CPP_WRAPPER=1``

> * For NHWC convolutions workloads

Reviewer: ```suggestion
* Convolution workloads may see a performance benefit by specifying  
```

>   ``torch._inductor.config.layout_optimization=True`` or ``TORCHINDUCTOR_LAYOUT_OPTIMIZATION=1``

Reviewer: ```suggestion
  ``torch._inductor.config.layout_optimization=True`` or ``TORCHINDUCTOR_LAYOUT_OPTIMIZATION=1``.
```

>   can help be enforcing channels_last format throughout the graph avoiding

Reviewer: ```suggestion
  This can help performance by enforcing channel_last memory format on the convolution in TorchInductor, avoiding any unnecessary transpose operations. 
```

>   any additional transposes added by ``inductor``. Note that

Reviewer: ```suggestion
  Note that
```

  ``PYTORCH_MIOPEN_SUGGEST_NHWC=1`` is recommended if using this.

* Extracting the Triton kernel ``TORCH_COMPILE_DEBUG`` creates a
  ``torch_compile_debug/`` directory at current path, in the ``output_code.py``
  the code-strings for the Triton kernels that are defined. Manual work is
  then required to strip out the kernel and create kernel

>   compilation and launch via Triton.

Reviewer: Can you change this whole paragraph to the following:
* To extract the triton kernels generated by inductor, set the environment variable ``TORCH_COMPILE_DEBUG=1``  which will create a ``torch_compile_debug/`` directory in the current path. The wrapper codes generated by inductor are in one or more ``output_code.py`` files corresponding to the FX graphs associated with model. The triton kernels are defined in these generated codes.

below:

.. code-block:: shell

   python -m vllm.entrypoints.api_server --model /path/to/model --dtype
   float16 -tp 2 --port 8000 &

To achieve both latency and throughput performance for serving, you can
run multiple API servers on different GPUs by specifying different ports
for each server and use ``ROCR_VISIBLE_DEVICES`` to specify the GPUs for
each server, for example:

.. code-block:: shell

   ROCR_VISIBLE_DEVICES=0,1 python -m vllm.entrypoints.api_server --model
   /path/to/model --dtype float16 -tp 2 --port 8000 &

   ROCR_VISIBLE_DEVICES=2,3 python -m vllm.entrypoints.api_server --model
   /path/to/model --dtype float16 -tp 2 --port 8001 &

Choose different attention backends
-----------------------------------

vLLM on ROCm supports three different attention backends:

-  **Triton Flash Attention** - For benchmarking, run vLLM scripts at
   least once as a warm-up step so Triton can perform auto-tuning before
   collecting benchmarking numbers. This is the default setting.

-  **Composable Kernel Flash Attention** - To use CK flash-attention, specify
   the environment variable as ``export VLLM_USE_TRITON_FLASH_ATTN=0``.

> -  **PyTorch naive attention** *-* To use naive attention, either build

Reviewer: ```suggestion
-  **PyTorch naive attention** *-* To use naive attention (PyTorch sdpa math backend), either build
```

throughput. Chunked prefill allows large prefills to be chunked into
smaller chunks and batched together with decode requests.

You can enable the feature by specifying ``--enable-chunked-prefill`` in the
command line or setting ``enable_chunked_prefill=True`` in the LLM
constructor. 

As stated in `vLLM's documentation, <https://docs.vllm.ai/en/latest/models/performance.html#chunked-prefill>`__,
you can tune the performance by changing ``max_num_batched_tokens``. By
default, it is set to 512 and optimized for ITL (inter-token latency).
Smaller ``max_num_batched_tokens`` achieves better ITL because there are
fewer prefills interrupting decodes.
Higher ``max_num_batched_tokens`` achieves better TTFT (time to the first
token) as you can put more prefill to the batch.

You might experience noticeable throughput improvements when
benchmarking on a single GPU or 8 GPUs using the vLLM throughput
benchmarking script along with the SharedGPT dataset as input.

In the case of fixed ``input-len``/``output-len``, for some configurations,
enabling chunked prefill increases the throughput. For some other
configurations, the throughput may be worse and elicit a need to tune

> ``parameter max_num_batched_tokens``.

Reviewer: ```suggestion
parameter ``max_num_batched_tokens`` (for example, increasing ``max_num_batched_tokens`` value to 4096 or larger).
```

To use ``ray``, specify the ``--worker-use-ray`` flag. The following script
example uses ``torchrun`` to run latency benchmarking using ``ray``
for ``input-len`` of 512, ``output-len`` of 512, and ``batch-size`` of 1:

.. code-block:: shell

   tp=$1

   torchrun --standalone --nnodes=1 --nproc-per-node=$tp benchmarks/benchmark_latency.py --worker-use-ray --model $MODEL --batch-size 1 --input-len 512 --output-len 512 --tensor-parallel-size $tp --num-iters 10

The first parameter of the script ``tp`` specifies the ``tensor-parallel`` size
(1 to 8).

GEMM tuning steps
^^^^^^^^^^^^^^^^^

1. Set various environment variables:

   .. code-block:: shell

      export VLLM_UNTUNE_FILE="/tmp/vllm_untuned.csv"

      export VLLM_TUNE_FILE="$(pwd)/vllm/tuned.csv"

      export HIP_FORCE_DEV_KERNARG=1

      export DEBUG_CLR_GRAPH_PACKET_CAPTURE=1

2. Do a tuning run:

   .. code-block:: shell

      VLLM_TUNE_GEMM=1 torchrun --standalone --nnodes=1 --nproc-per-node=8 vllm/benchmarks/benchmark_latency.py --batch-size 1 --input-len 2048 --output-len 128 --model /models/llama-2-70b-chat-hf/ -tp 8

>       python vllm/gradlib/gemm_tuner.py --input /tmp/vllm_untuned.csv --tuned_file vllm/tuned.csv

Reviewer: ```suggestion
      python $PATH_TO_GRADLIB/gemm_tuner.py --input /tmp/vllm_untuned.csv --tuned_file vllm/tuned.csv
```

   .. note::

>       It seems the gradlib is no longer in vllm directory. To find

Reviewer: ```suggestion
      where ``$PATH_TO_GRADLIB`` is the installation path of ``gradlib``. To find
```

>    .. note::

Reviewer: ```suggestion
 
```

Remove "note" here.
